cmd/kube-controller-manager/app/core_test.go

```diff
@@ -141,3 +143,175 @@ func TestController_DiscoveryError(t *testing.T) {
 		}
 	}
 }
+
+func TestSetNodeCIDRMaskSizes(t *testing.T) {
+	_, dummyIPv4CIDR, err := net.ParseCIDR("1.2.3.0/32")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, dummyIPv6CIDR, err := net.ParseCIDR("2001:db1::/64")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tcs := map[string]struct {
+		clusterCIDRs     []*net.IPNet
+		cfg              nodeipamconfig.NodeIPAMControllerConfiguration
+		expectedIPv4Mask int
+		expectedIPv6Mask int
+		expectedErr      bool
+	}{
+		"dual-stack: set NodeCIDRMaskSize": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+				dummyIPv6CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSize: 28,
+			},
+			expectedIPv4Mask: defaultNodeMaskCIDRIPv4,
+			expectedIPv6Mask: defaultNodeMaskCIDRIPv6,
+			expectedErr:      true,
+		},
+		"dual-stack: set NodeCIDRMaskSizeIPv4 only": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+				dummyIPv6CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSizeIPv4: 28,
+			},
+			expectedIPv4Mask: 28,
+			expectedIPv6Mask: defaultNodeMaskCIDRIPv6,
+			expectedErr:      false,
+		},
+		"dual-stack: set NodeCIDRMaskSizeIPv6 only": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+				dummyIPv6CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSizeIPv6: 78,
+			},
+			expectedIPv4Mask: defaultNodeMaskCIDRIPv4,
+			expectedIPv6Mask: 78,
+			expectedErr:      false,
+		},
+		"dual-stack: set both NodeCIDRMaskSizeIPv4 and NodeCIDRMaskSizeIPv6": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+				dummyIPv6CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSizeIPv4: 28,
+				NodeCIDRMaskSizeIPv6: 78,
+			},
+			expectedIPv4Mask: 28,
+			expectedIPv6Mask: 78,
+			expectedErr:      false,
+		},
+		"single-stack IPv4: set NodeCIDRMaskSizeIPv6": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSizeIPv6: 78,
+			},
+			expectedIPv4Mask: defaultNodeMaskCIDRIPv4,
+			expectedIPv6Mask: defaultNodeMaskCIDRIPv6,
+			expectedErr:      true,
+		},
+		"single-stack IPv4: set NodeCIDRMaskSize": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSize: 28,
+			},
+			expectedIPv4Mask: 28,
+			expectedIPv6Mask: defaultNodeMaskCIDRIPv6,
+			expectedErr:      false,
+		},
+		"single-stack IPv4: set NodeCIDRMaskSizeIPv4": {
+			clusterCIDRs: []*net.IPNet{
+				dummyIPv4CIDR,
+			},
+			cfg: nodeipamconfig.NodeIPAMControllerConfiguration{
+				NodeCIDRMaskSizeIPv4: 28,
+			},
+			expectedIPv4Mask: 28,
+			expectedIPv6Mask: defaultNodeMaskCIDRIPv6,
+			expectedErr:      false,
+		},
+		"single-stack IPv4: set both NodeCIDRMaskSize and NodeCIDRMaskSizeIPv4": {
```

Comment:
@danwinship is this something that you had in mind, that `--node-cidr-mask-size-ipv4` takes precedence over `--node-cidr-mask-size`?
I don't know, I think it's changing the current behavior :/

Reply:
@aojea Not sure how to deal with them.
`node-cidr-mask-size-ipv4` seems more accurate, so it will override `node-cidr-mask-size`.
If the user provides `node-cidr-mask-size-ipv4` flag, he may want to set IPv4Mask as that value.

Reply:
agree with accuracy, but we are overriding one flag behavior ... just don't know if this is the correct way to do it

Reply:
I would say that it should be an error to specify both. It should certainly be an error to specify both with different values.